jax/util.py

```diff
@@ -242,3 +244,69 @@ def wrap_name(name, transform_name):
 
 def extend_name_stack(stack, name=''):
   return stack + name + '/'
+
+
+class ContextVar:
+  """Like contextvars.ContextVar, but implemented with threading.local()."""
+  # TODO(shoyer): remove in favor of contextvars when JAX requires Python 3.7+
+
+  def __init__(self, name, *, default):
+    self.name = name
+    self.default = default
+    self._state = threading.local()
+    self._state.value = default
+    self._lock = threading.Lock()
+
+  def get(self):
+    return getattr(self._state, 'value', self.default)
+
+  def set(self, value):
+    with self._lock:
+      old_value = self.get()
+      self._state.value = value
+    return old_value
+
+  def reset(self, token):
+    self._state.value = token
+
+
+_OVERRIDES = {}
+
+
+def overrideable(name: str):
+  """Make an internal JAX function overrideable."""
+  def decorator(fun: Callable):
+    _OVERRIDES[name] = ContextVar(name, default=fun)
+    @functools.wraps(fun)
+    def wrapper(*args, **kwargs):
+      impl = _OVERRIDES[name].get()
+      with override_context({name: fun}):
+        return impl(*args, **kwargs)
+    return wrapper
+  return decorator
+
+@contextlib.contextmanager
+def override_context(implementations: Mapping[str, Callable]):
+  """Override the implementations of JAX functions within a context.
+
+  This is function is EXPERIMENTAL and offers no backwards compatibility
+  guarantees. It exists for the sole purpose of overriding the implementation of
+  higher order JAX functions within a limited scope, particularly for libraries
+  such as Haiku and Flax that implement their own versions of this functions
+  that support mutation.
+
+  Usage example:
+
+    with jax.override_context({'lax.scan': my_scan}):
+      # all calls to lax.scan() are replaced by my_scan()
```

Comment:
Maybe it is worth documenting here that all cals to `lax.scan` inside `my_scan` will use the version of `lax.scan` before the override.